### PR TITLE
Add:Sparse Transfer Learning(STL) Recipe + Tutorial

### PR DESCRIPTION
**Pull request by rahul-tuli**
* Resnet
* MobileNet

### CODE REVIEW

File: integrations/pytorch/recipes/sparse-transfer-learn.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+
+---
+# Epoch and Learning-Rate variables
+num_epochs: &num_epochs 10.0
```
Comment (kevinaer) [inline integrations/pytorch/recipes/sparse-transfer-learn.md:19]: Should we be including this in a code block?

File: integrations/pytorch/recipes/sparse-transfer-learn.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  - !LearningRateModifier
+    start_epoch: 0.0
+    lr_class: MultiStepLR
+    lr_kwargs:
+      milestones: *lr_step_milestones
+      gamma: 0.1
+    init_lr: *init_lr
+
+# Phase 1 Sparse Transfer Learning / Recovery
+sparse_transfer_learning_modifiers:
```
Comment (kevinaer) [inline integrations/pytorch/recipes/sparse-transfer-learn.md:38]: Same as above regarding code block

File: integrations/pytorch/recipes/sparse-transfer-learn.md (hunk 1 above)
Comment (bfineran) [reply integrations/pytorch/recipes/sparse-transfer-learn.md:19]: this is a recipe, so the YAML portion is in the front matter (enclosed by `---`)

File: integrations/pytorch/recipes/sparse-transfer-learn.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+  - !ConstantPruningModifier
+    start_epoch: 0.0
+    params: __ALL__
+    
+---
+
+# Transfer Learning
+
+This recipe provides a framework for performing sparse transfer learning on
+SparseML's `sparsified` ResNet50, or MobileNet model.  This recipe was written for transfer learning
```
Comment (jeanniefinks) [inline integrations/pytorch/recipes/sparse-transfer-learn.md:48]:
```suggestion
SparseML's `sparsified` ResNet-50, or MobileNet model.  This recipe was written for transfer learning
```

File: integrations/pytorch/recipes/sparse-transfer-learn.md [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+onto the [Imagenette](https://github.com/fastai/imagenette) dataset using
+a single GPU, Adam optimizer, and batch size 32.  Adjust the hyperparameters at the top of the file accordingly
+for your dataset and training environment.
+
+## Imagenette
+[Imagenette](https://github.com/fastai/imagenette) is a subset of 10 easily classified classes from Imagenet (tench, English springer, cassette player, chain saw, church, French horn, garbage truck, gas pump, golf ball, parachute).
```
Comment (jeanniefinks) [inline integrations/pytorch/recipes/sparse-transfer-learn.md:54]:
```suggestion
[Imagenette](https://github.com/fastai/imagenette) is a subset of 10 easily classified classes from ImageNet (tench, English springer, cassette player, chain saw, church, French horn, garbage truck, gas pump, golf ball, parachute).
```

File: integrations/pytorch/recipes/sparse-transfer-learn.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+To transfer learn Resnet50 on [Imagenette](https://github.com/fastai/imagenette) with this recipe, run the following example command.
```
Comment (jeanniefinks) [inline integrations/pytorch/recipes/sparse-transfer-learn.md:55]:
```suggestion
To transfer learn ResNet-50 on [Imagenette](https://github.com/fastai/imagenette) with this recipe, run the following example command.
```

File: integrations/pytorch/recipes/sparse-transfer-learn.md (hunk 1 above)
Comment (rahul-tuli) [reply integrations/pytorch/recipes/sparse-transfer-learn.md:19]: Will leave it as is for now. Will agree the codeblock makes rendered md look way nicer.

File: integrations/pytorch/recipes/sparse-transfer-learn.md (hunk 2 above)
Comment (rahul-tuli) [reply integrations/pytorch/recipes/sparse-transfer-learn.md:38]: Same as above!

**Review by kevinaer — DISMISSED** [review]

File: integrations/pytorch/tutorials/sparse_transfer_learning_tutorial.md [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+  - !ConstantPruningModifier
+    start_epoch: 0.0
+    params: __ALL__
+    
+---
+```
+
+`ConstantPruningModifier` maintains model's sparsity levels during transfer learning process.
+We include this example [recipe](https://github.com/neuralmagic/sparseml/blob/main/integrations/pytorch/recipes/sparse-transfer-learn.md) in the SparseML [GitHub repository](https://github.com/neuralmagic/sparseml).
+To learn more about recipes and modifiers refer [here](https://github.com/neuralmagic/sparseml/blob/main/docs/source/recipes.md)
```
Comment (jeanniefinks) [inline integrations/pytorch/tutorials/sparse_transfer_learning_tutorial.md:114]:
```suggestion
[Learn more about recipes and modifiers](https://github.com/neuralmagic/sparseml/blob/main/docs/source/recipes.md).
```

File: integrations/pytorch/tutorials/sparse_transfer_learning_tutorial.md [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+The script automatically saves model checkpoints after each epoch and reports the validation loss along with layer sparsities. 
+The model is saved in [ONNX](https://onnx.ai/) format and can be loaded later for inference or other experiments.
+
+## Wrap-Up
+
+Neural Magic sparse models and recipes simplify the sparsification process by enabling sparse transfer learning to create highly accurate pruned image classification models. 
+In this tutorial, you downloaded a pre-sparsified model, applied a Neural Magic recipe for sparse transfer learning, and saved the model for future use.
+
+Now, refer [here](https://github.com/neuralmagic/deepsparse/tree/main/examples/classification) for an example for benchmarking and deploying image classification models with DeepSparse.
```
Comment (jeanniefinks) [inline integrations/pytorch/tutorials/sparse_transfer_learning_tutorial.md:157]:
```suggestion
 An example for benchmarking and deploying image classification models with DeepSparse [is also available](https://github.com/neuralmagic/deepsparse/tree/main/examples/classification).
```

**Review by jeanniefinks — CHANGES_REQUESTED** [review]

**Comment by rahul-tuli** [comment]
Thank you @jeanniefinks!

**Review by jeanniefinks — DISMISSED** [review]
🚀

File: integrations/pytorch/recipes/sparse-transfer-learn.md [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    lr_kwargs:
+      milestones: *lr_step_milestones
+      gamma: 0.1
+    init_lr: *init_lr
+
+# Phase 1 Sparse Transfer Learning / Recovery
+sparse_transfer_learning_modifiers:
+  - !ConstantPruningModifier
+    start_epoch: 0.0
+    params: __ALL__
```
Comment (bfineran) [inline integrations/pytorch/recipes/sparse-transfer-learn.md:41]: this will only grab the conv/FC weights and ignore stuff like batch norm

```suggestion
    params: __ALL_PRUNABLE__
```

File: integrations/pytorch/recipes/sparse-transfer-learn.md [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+This recipe provides a framework for performing sparse transfer learning on
+SparseML's `sparsified` ResNet-50, or MobileNet model.  This recipe was written for transfer learning
+onto the [Imagenette](https://github.com/fastai/imagenette) dataset using
+a single GPU, Adam optimizer, and batch size 32.  Adjust the hyperparameters at the top of the file accordingly
+for your dataset and training environment.
+
+## Imagenette
+[Imagenette](https://github.com/fastai/imagenette) is a subset of 10 easily classified classes from ImageNet (tench, English springer, cassette player, chain saw, church, French horn, garbage truck, gas pump, golf ball, parachute).
+To transfer learn ResNet-50 on [Imagenette](https://github.com/fastai/imagenette) with this recipe, run the following example command.
```
Comment (bfineran) [inline integrations/pytorch/recipes/sparse-transfer-learn.md:55]:
```suggestion


To transfer learn ResNet-50 on [Imagenette](https://github.com/fastai/imagenette) with this recipe, run the following example command.
```

File: integrations/pytorch/recipes/sparse-transfer-learn.md [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+    --model-tag mobilenet-imagenette-sparse-transfer-learned
+```
+## Other Datasets
+To transfer learn this sparsified model to other datasets
+you may have to adjust certain hyperparameters in this recipe and/or training script.
+Some considerations:
+* For more complex datasets, increate the number of epochs, adjusting the learning rate step accordingly
+* Adding more learning rate step milestones can lead to more jumps in accuracy
+* Increasing the number of epochs before starting quantization can give a higher baseline
+* If accuracy is not recovering in the quantization phase, the total number of epochs should be increased as well
```
Comment (bfineran) [inline integrations/pytorch/recipes/sparse-transfer-learn.md:95]: nit: remove if recipe does not include quantization

File: integrations/pytorch/tutorials/sparse_transfer_learning_tutorial.md [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+# Sparse Transfer Learning for Image Classification
+
+This tutorial shows how Neural Magic sparse models simplify the sparsification process by offering pre-sparsified models for transfer learning onto other datasets.
+
+## Overview
+
+Neural Magic’s ML team creates sparsified models that allow anyone to plug in their data and leverage pre-sparsified models from the SparseZoo. 
+Sparsifying involves removing redundant information from neural networks using algorithms such as pruning and quantization, among others. 
+This sparsification process results in many benefits for deployment environments, including faster inference and smaller file sizes. 
+Unfortunately, many have not realized the benefits due to the complicated process and number of hyperparameters involved.
```
Comment (bfineran) [inline integrations/pytorch/tutorials/sparse_transfer_learning_tutorial.md:26]:
```suggestion

Unfortunately, many have not realized the benefits due to the complicated process and number of hyperparameters involved.
```

**Review by bfineran — DISMISSED** [review]
Good work @rahul-tuli! Left a few suggestions. The only procedural thing that could be changed is that the `vision.py` script can actually take SparseZoo stubs as a checkpoint input, but I think it's also nice to have the weight download as part of the tutorial...

**Review by bfineran — COMMENTED** [review]
one more thing, let's remove the quantized and base models from the table of SparseZoo models since for the recipe to work, the user will specifically need to use a FP32 sparse model checkpoint

File: integrations/pytorch/recipes/sparse-transfer-learn.md (hunk 10 above)
Comment (rahul-tuli) [reply integrations/pytorch/recipes/sparse-transfer-learn.md]: Eye of the tiger! 🐅

**Review by bfineran — DISMISSED** [review]

**Review by jeanniefinks — DISMISSED** [review]

**Review by lucaslie — DISMISSED** [review]

**Review by bfineran — DISMISSED** [review]

File: integrations/pytorch/recipes/sparse-transfer-learn.md [hunk 12]
```diff
@@ -0,0 +1,98 @@
+<!--
```
Comment (markurtz) [inline integrations/pytorch/recipes/sparse-transfer-learn.md:1]: can we rename this to follow standards newly setup in the yolov5 integration? Specifically [model].transfer_learn_pruned.md and for model let's mark it as generic for classification by putting in 'classification' in place

File: integrations/pytorch/tutorials/sparse_transfer_learning_tutorial.md [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+
+This tutorial can be run by cloning and installing the `sparseml` repository which contains scripts and recipes for
+this example:
+
+```bash
+git clone https://github.com/neuralmagic/sparseml.git
+pip install sparseml[torchvision]
+```
+Note:  make sure to upgrade `pip` using `python -m pip install -U pip`
+## Downloading and Preparing a Pre-Sparsified Model
```
Comment (markurtz) [inline integrations/pytorch/tutorials/sparse_transfer_learning_tutorial.md:48]: We should be able to get by without having the user run the SparseZoo code to download. Specifically the vision.py script should accept the stubs to start from and auto download from the zoo, see the YOLOv5 tutorial as an example

File: integrations/pytorch/tutorials/sparse_transfer_learning_tutorial.md [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+model = Zoo.load_model_from_stub(stub)
+downloded_path = model.framework_files[-1].downloaded_path()
+print(f'model with stub {stub} downloaded to {downloded_path}')
+```
+Note the model checkpoint path; this will be used to load the sparsified model.
+
+## Applying a Sparse Transfer Learning Recipe
+
+- Once the model checkpoint has been downloaded, [vision.py](https://github.com/neuralmagic/sparseml/blob/main/integrations/pytorch/vision.py) script can be used to download [Imagenette](https://github.com/fastai/imagenette) and kick-start transfer learning. 
+   The transfer learning process itself is guided using recipes; a minimal example recipe could look like the following:
```
Comment (markurtz) [inline integrations/pytorch/tutorials/sparse_transfer_learning_tutorial.md:78]: I'm not sure if we should include an example recipe in here. If they want to learn more about what the recipes look like, then we should have an external tutorial and the external recipes should contain enough information. This is going to additionally compound once we add in the quantization sparse transfer learning

**Review by markurtz — CHANGES_REQUESTED** [review]
A few comments. Additionally for the overall structure we should follow the numbered step setup within each sub category like what was done for v3 and v5 tutorials

File: integrations/pytorch/recipes/sparse-transfer-learn.md (hunk 12 above)
Comment (rahul-tuli) [reply integrations/pytorch/recipes/sparse-transfer-learn.md]: Resolved!

File: integrations/pytorch/tutorials/sparse_transfer_learning_tutorial.md (hunk 13 above)
Comment (rahul-tuli) [reply integrations/pytorch/tutorials/sparse_transfer_learning_tutorial.md]: Resolved!

File: integrations/pytorch/tutorials/sparse_transfer_learning_tutorial.md (hunk 14 above)
Comment (rahul-tuli) [reply integrations/pytorch/tutorials/sparse_transfer_learning_tutorial.md]: Removed!

Comment (rahul-tuli) [reply integrations/pytorch/tutorials/sparse_transfer_learning_tutorial.md]: Okay, removed recipes from tuotrial!

**Review by markurtz — APPROVED** [review]

**Review by bfineran — APPROVED** [review]

**Comment by rahul-tuli** [comment]
Thank you Sensei @markurtz @bfineran
